src/server/pfs/server/driver_new_storage.go

```diff
@@ -85,45 +83,67 @@ func (d *driver) finishCommitNewStorageLayer(txnCtx *txnenv.TransactionContext,
 	return d.writeFinishedCommit(txnCtx.Stm, commit, commitInfo)
 }
 
-// (bryce) should have a context going through here.
-func (d *driver) putFilesNewStorageLayer(pachClient *client.APIClient, server *putFileServer) error {
+func (d *driver) putFilesNewStorageLayer(pachClient *client.APIClient, server *putFileServer) (retErr error) {
 	// (bryce) how to handle no commit started?
-	// (bryce) this will not work for large files with the current implementation.
-	for req, err := server.Recv(); err == nil; req, err = server.Recv() {
-		if err := d.fs.WriteHeader(&tar.Header{Name: req.File.Path}); err != nil {
-			return err
-		}
-		if _, err := io.Copy(d.fs, bytes.NewReader(req.Value)); err != nil {
-			return err
+	// (bryce) a failed put (crash/error) should result with any serialized sub file sets getting
+	// cleaned up.
+	// (bryce) piping is not ideal, but easy for getting integration started.
+	// we will want to eventually just read raw bytes from the grpc stream.
+	r, w := io.Pipe()
+	var eg errgroup.Group
+	var fs *fileset.FileSet
+	fsChan := make(chan *fileset.FileSet)
```

Comment:
I'm a little confused how this channel is being used, it looks like there's only one recv call on the channel (on line 97) but there can be many sends to the channel from line 112. Am I misreading this, or is there a potential deadlock here? It seems like there is, even if there isn't it seems like this code could maybe be simplified a bit to not looks like there are.

Reply:
This is refactored. To answer your question though, there would only be one send to fsChan for the file set created in the rpc, so I don't think this can deadlock. The goroutine that creates the fileset would block on the write until the read end of the pipe was read by the other goroutine (which sets fs before reading). This is definitely not the approach I wanted to take as mentioned in the comment, but I think we are good now with the new pfs api endpoints.